File: apple/Sources/CCDAEngine/Models/CCDATimestamp.swift
import Foundation

/// HL7 timestamp value parsed into date and time components.
///
/// Supports C-CDA/HL7 TS literals in these forms:
///
/// ```text
/// YYYY
/// YYYYMM
/// YYYYMMDD
/// YYYYMMDDHH
/// YYYYMMDDHHMM
/// YYYYMMDDHHMMSS
/// YYYYMMDDHHMMSS.fraction
/// YYYYMMDDHHMMSS+ZZZZ
/// YYYYMMDDHHMMSS-ZZZZ
/// YYYYMMDDHHMMSS.fraction+ZZZZ
/// YYYYMMDDHHMMSS.fraction-ZZZZ
/// ```
///
/// Missing components remain `nil`, so a date-only value such as `19800515`
/// does not invent an hour, minute, or second.
public struct CCDATimestamp: Hashable, Sendable {
    /// Original timestamp value exactly as provided by the XML.
    public let rawValue: String
    /// Four-digit year component.
    public let year: Int
    /// Month component from 1 through 12.
    public let month: Int?
    /// Day component from 1 through 31, validated when month and year are available.
    public let day: Int?
    /// Hour component from 0 through 23.
    public let hour: Int?
    /// Minute component from 0 through 59.
    public let minute: Int?
    /// Second component from 0 through 59.
    public let second: Int?
    /// Fractional second text without the leading decimal point.
    public let fractionalSecond: String?
    /// Time zone offset text such as `-0500`, when provided.
    public let timeZoneOffset: String?

    /// Whether the timestamp includes any time-of-day component.
    public var hasTime: Bool { hour != nil }

    /// Creates a timestamp by parsing an HL7 TS literal.
    public init?(rawValue: String) {
        self.rawValue = rawValue

        let trimmedValue = rawValue.trimmingCharacters(in: .whitespacesAndNewlines)
        let splitValue = Self.splitTimeZone(from: trimmedValue)
        let splitFraction = Self.splitFraction(from: splitValue.timestamp)
        let digits = splitFraction.timestamp

        guard Self.isSupportedDigitCount(digits.count),
              digits.allSatisfy(\.isNumber),
              let parsedYear = Self.integer(from: digits, start: 0, length: 4),
              (1...9999).contains(parsedYear) else {
            return nil
        }

        let parsedMonth = Self.integer(from: digits, start: 4, length: 2)
        let parsedDay = Self.integer(from: digits, start: 6, length: 2)
        let parsedHour = Self.integer(from: digits, start: 8, length: 2)
        let parsedMinute = Self.integer(from: digits, start: 10, length: 2)
        let parsedSecond = Self.integer(from: digits, start: 12, length: 2)

        guard Self.optionalComponentsAreValid(
            month: parsedMonth,
            day: parsedDay,
            hour: parsedHour,
            minute: parsedMinute,
            second: parsedSecond
        ) else {
            return nil
        }

        self.year = parsedYear
        self.month = parsedMonth
        self.day = parsedDay
        self.hour = parsedHour
        self.minute = parsedMinute
        self.second = parsedSecond
        self.fractionalSecond = splitFraction.fractionalSecond
        self.timeZoneOffset = splitValue.timeZoneOffset
    }

    private static func splitTimeZone(from value: String) -> (timestamp: String, timeZoneOffset: String?) {
        let splitIndex = value.lastIndex { $0 == "+" || $0 == "-" }
        guard let splitIndex else { return (value, nil) }

        let timestamp = String(value[..<splitIndex])
        let offset = String(value[splitIndex...])

        guard offset.count == 5,
              offset.dropFirst().allSatisfy(\.isNumber) else {
            return (value, nil)
        }

        return (timestamp, offset)
    }

    private static func splitFraction(from value: String) -> (timestamp: String, fractionalSecond: String?) {
        let parts = value.split(separator: ".", maxSplits: 1, omittingEmptySubsequences: false)

        guard parts.count == 2,
              !parts[1].isEmpty,
              parts[1].allSatisfy(\.isNumber) else {
            return (value, nil)
        }

        return (String(parts[0]), String(parts[1]))
    }

    private static func isSupportedDigitCount(_ count: Int) -> Bool {
        [4, 6, 8, 10, 12, 14].contains(count)
    }

    private static func integer(from value: String, start: Int, length: Int) -> Int? {
        guard value.count >= start + length else {
            return nil
        }

        let startIndex = value.index(value.startIndex, offsetBy: start)
        let endIndex = value.index(startIndex, offsetBy: length)
        return Int(value[startIndex..<endIndex])
    }

    private static func optionalComponentsAreValid(
        month: Int?,
        day: Int?,
        hour: Int?,
        minute: Int?,
        second: Int?
    ) -> Bool {
        if let month, !(1...12).contains(month) {
            return false
        }

        if let day, !(1...31).contains(day) {
            return false
        }

        if let hour, !(0...23).contains(hour) {
            return false
        }

        if let minute, !(0...59).contains(minute) {
            return false
        }

        if let second, !(0...59).contains(second) {
            return false
        }
        return true
    }
}

File: android/ccda-engine/src/main/kotlin/com/ccdakit/engine/Models/Common/CCDATimestamp.kt
package com.ccdakit.engine

/**
 * Parsed HL7 TS timestamp preserving the original precision instead of forcing every value into a date-time.
 *
 * Supports `yyyy`, `yyyyMM`, `yyyyMMdd`, `yyyyMMddHH`, `yyyyMMddHHmm`,
 * and `yyyyMMddHHmmss`, with optional `.fraction` and `+/-ZZZZ` timezone offset.
 *
 * @property rawValue Original timestamp string.
 * @property year Parsed year.
 * @property month Parsed month when present.
 * @property day Parsed day when present.
 * @property hour Parsed hour when present.
 * @property minute Parsed minute when present.
 * @property second Parsed second when present.
 * @property fractionalSecond Fractional second digits when present.
 * @property timeZoneOffset Timezone offset such as `-0500` when present.
 */
data class CCDATimestamp(
    val rawValue: String,
    val year: Int,
    val month: Int? = null,
    val day: Int? = null,
    val hour: Int? = null,
    val minute: Int? = null,
    val second: Int? = null,
    val fractionalSecond: String? = null,
    val timeZoneOffset: String? = null,
) {
    /** Whether the source timestamp included an hour component. */
    val hasTime: Boolean
        get() = hour != null

    companion object {
        private val supportedDigitCounts = setOf(4, 6, 8, 10, 12, 14)

        /** Parses an HL7 TS timestamp into positional date/time components. */
        fun parse(rawValue: String?): CCDATimestamp? {
            val raw = rawValue?.trim()?.takeIf { it.isNotEmpty() } ?: return null
            val (timestampWithFraction, offset) = splitTimeZone(raw)
            val (digits, fraction) = splitFraction(timestampWithFraction)

            if (digits.length !in supportedDigitCounts || !digits.all { it.isDigit() }) {
                return null
            }

            val year = digits.intAt(0, 4)?.takeIf { it in 1..9999 } ?: return null
            val month = digits.intAt(4, 2)
            val day = digits.intAt(6, 2)
            val hour = digits.intAt(8, 2)
            val minute = digits.intAt(10, 2)
            val second = digits.intAt(12, 2)

            if (!optionalComponentsAreValid(month, day, hour, minute, second)) {
                return null
            }

            return CCDATimestamp(
                rawValue = raw,
                year = year,
                month = month,
                day = day,
                hour = hour,
                minute = minute,
                second = second,
                fractionalSecond = fraction,
                timeZoneOffset = offset,
            )
        }

        private fun splitTimeZone(value: String): Pair<String, String?> {
            val index = value.indexOfLast { it == '+' || it == '-' }
            if (index < 0) return value to null

            val timestamp = value.substring(0, index)
            val offset = value.substring(index)

            return if (offset.length == 5 && offset.drop(1).all { it.isDigit() }) {
                timestamp to offset
            } else {
                value to null
            }
        }

        private fun splitFraction(value: String): Pair<String, String?> {
            val parts = value.split(".", limit = 2)
            if (parts.size != 2 || parts[1].isEmpty() || !parts[1].all { it.isDigit() }) {
                return value to null
            }
            return parts[0] to parts[1]
        }

        private fun String.intAt(start: Int, length: Int): Int? {
            if (this.length < start + length) return null
            return substring(start, start + length).toIntOrNull()
        }

        private fun optionalComponentsAreValid(
            month: Int?,
            day: Int?,
            hour: Int?,
            minute: Int?,
            second: Int?,
        ): Boolean {
            if (month != null && month !in 1..12) return false
            if (day != null && day !in 1..31) return false
            if (hour != null && hour !in 0..23) return false
            if (minute != null && minute !in 0..59) return false
            if (second != null && second !in 0..59) return false
            return true
        }
    }
}
